File: elixir/roman-numerals/lib/roman_numerals.ex
defmodule RomanNumerals do
  @roman_letters [{"I", "V"}, {"X", "L"}, {"C", "D"}, {"M", "?"}]
  @doc """
  Convert the number to a roman number.
  """
  @spec numeral(pos_integer) :: String.t()
  def numeral(number) do
    if number < 0 || number >= 4000, do: raise "no conversion to roman for this number"
    Integer.digits(number)
    |> Enum.reverse()
    |> Enum.with_index()
    |> Enum.reduce("", fn el, acc -> get_number(el) <> acc end)
  end

  defp get_number({digit, pos}) do
    {a, b} = @roman_letters |> Enum.at(pos)
    case digit do
      n when n in [0, 1, 2, 3] -> String.duplicate(a, n)
      4 -> a <> b
      r when r in [5, 6, 7, 8] -> b <> String.duplicate(a, r - 5)
      9 -> a <> (@roman_letters |> Enum.at(pos + 1) |> elem(0))
      _ -> raise "not a digit"
    end
  end
end

File: fsharp/roman-numerals/RomanNumerals.fs
module RomanNumerals

let roman arabicNumeral = 
    if arabicNumeral >= 4000 then failwith "Number too big"
    let digits = arabicNumeral |> string |> Seq.map (int >> (+) (- int '0')) |> Seq.rev |> Seq.indexed
    let romanLetters = [("I", "V"); ("X", "L"); ("C", "D"); ("M", "?")]
    let getRightNumber (pos, digit) = 
        let (a, b) = romanLetters.[pos]
        match digit with 
        | 0 -> ""
        | 1 -> a 
        | 2 -> a + a 
        | 3 -> a + a + a 
        | 4 -> a + b 
        | 5 -> b 
        | 6 -> b + a 
        | 7 -> b + a + a 
        | 8 -> b + a + a + a 
        | 9 -> a + (fst (romanLetters.[pos + 1]))
        | _ -> failwith "not a digit"
    Seq.foldBack (fun el acc -> acc + getRightNumber el) digits ""